sdk/src/types/block/address/bech32.rs

```diff
@@ -1,17 +1,180 @@
 // Copyright 2023 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-use alloc::string::String;
+use alloc::{
+    string::{String, ToString},
+    vec::Vec,
+};
 use core::str::FromStr;
 
+use bech32::{FromBase32, ToBase32, Variant};
 use derive_more::{AsRef, Deref};
+use packable::{error::UnpackError, packer::Packer, unpacker::Unpacker, Packable, PackableExt};
 
 use crate::types::block::{address::Address, Error};
 
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct Hrp {
+    inner: [u8; 83],
+    len: u8,
+}
+
+impl Hrp {
+    /// Convert a string to an Hrp without checking validity.
+    pub const fn from_str_unchecked(hrp: &str) -> Self {
+        let len = hrp.len();
+        let mut bytes = [0; 83];
+        let hrp = hrp.as_bytes();
+        let mut i = 0;
+        while i < len {
+            bytes[i] = hrp[i];
+            i += 1;
+        }
+        Self {
+            inner: bytes,
+            len: len as _,
+        }
+    }
+}
+
+impl FromStr for Hrp {
+    type Err = Error;
+
+    fn from_str(hrp: &str) -> Result<Self, Self::Err> {
+        let len = hrp.len();
+        if hrp.is_ascii() && len <= 83 {
+            let mut bytes = [0; 83];
+            bytes[..len].copy_from_slice(hrp.as_bytes());
+            Ok(Self {
+                inner: bytes,
+                len: len as _,
+            })
+        } else {
+            Err(Error::InvalidBech32Hrp(hrp.to_string()))
+        }
+    }
+}
+
+impl core::fmt::Display for Hrp {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        let hrp_str = self.inner[..self.len as usize]
+            .iter()
+            .map(|b| *b as char)
+            .collect::<String>();
+        f.write_str(&hrp_str)
+    }
+}
+
+impl Packable for Hrp {
+    type UnpackError = Error;
+    type UnpackVisitor = ();
+
+    #[inline]
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        self.len.pack(packer)?;
+        packer.pack_bytes(self.inner)?;
+
+        Ok(())
+    }
+
+    #[inline]
+    fn unpack<U: Unpacker, const VERIFY: bool>(
+        unpacker: &mut U,
+        visitor: &Self::UnpackVisitor,
+    ) -> Result<Self, UnpackError<Self::UnpackError, U::Error>> {
+        let Ok(len) = u8::unpack::<_, VERIFY>(unpacker, visitor) else { unreachable!() };
```

Comment:
Hmm, depending on the packer, this seems definitely reachable? Why not just propagating the error?

Reply:
The error is `Infallible`.

Reply:
Why not just `unwrap()`, or `.coerce()?` ?